compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64AtomicMove.java

```diff
@@ -56,34 +59,52 @@
     public static class CompareAndSwapOp extends AArch64LIRInstruction {
         public static final LIRInstructionClass<CompareAndSwapOp> TYPE = LIRInstructionClass.create(CompareAndSwapOp.class);
 
+        private final AArch64Kind accessKind;
+        private final int memoryBarrier;
+
         @Def protected AllocatableValue resultValue;
         @Alive protected Value expectedValue;
         @Alive protected AllocatableValue newValue;
         @Alive protected AllocatableValue addressValue;
         @Temp protected AllocatableValue scratchValue;
 
-        public CompareAndSwapOp(AllocatableValue result, Value expectedValue, AllocatableValue newValue, AllocatableValue addressValue, AllocatableValue scratch) {
+        public CompareAndSwapOp(AArch64Kind accessKind, AllocatableValue result, Value expectedValue, AllocatableValue newValue, AllocatableValue addressValue, AllocatableValue scratch,
+                        int memoryBarrier) {
             super(TYPE);
+            this.accessKind = accessKind;
             this.resultValue = result;
             this.expectedValue = expectedValue;
             this.newValue = newValue;
             this.addressValue = addressValue;
             this.scratchValue = scratch;
+            this.memoryBarrier = memoryBarrier;
         }
 
         @Override
         public void emitCode(CompilationResultBuilder crb, AArch64MacroAssembler masm) {
-            AArch64Kind kind = (AArch64Kind) expectedValue.getPlatformKind();
-            assert kind.isInteger();
-            final int size = kind.getSizeInBytes() * Byte.SIZE;
+            assert accessKind.isInteger();
+            final int size = accessKind.getSizeInBytes() * Byte.SIZE;
 
             Register address = asRegister(addressValue);
             Register result = asRegister(resultValue);
             Register newVal = asRegister(newValue);
+            Register expected = asRegister(expectedValue);
+
+            // For the atomics with "acquire/release" suffixes, the corresponding memory barrier has
+            // been inserted after/before the CAS respectively. Thus, the CAS operation here does
+            // not need the "acquire" semantics. Contrarily, the strong variants, i.e., atomics
+            // without "acquire/release" suffixes, are not surrounded by memory barriers. Thus, the
+            // CAS operation needs to use the variant of instruction that has the "acquire" effect.
+            boolean acquire = memoryBarrier >= (LOAD_LOAD | LOAD_STORE | STORE_STORE);
+
+            // The exclusive store is always using the "release" semantic. That is required to
+            // ensure visibility to other threads of the exclusive write (assuming it succeeds)
+            // before that of any subsequent writes.
+            boolean release = true;
```

Comment:
FYI a release only affects instructions before the memory access, not after it. Subsequent memory operations can be hoisted. W.r.t. another thread, an acquire is needed to be issued by the other thread as well in order to have any ordering guarantees.